src/calculateAverage.js

```diff
@@ -37,7 +37,32 @@ function applyCalculateAverage() {
    * @param {Object} options - optional
    */
   [].__proto__.calculateAverage = function(options) {
-    // write code here
+    if (options === undefined) {
+      return this.reduce((a, b) => (a + b)) / this.length;
+    }
+
+    if (options instanceof Object) {
+      if (options.hasOwnProperty('accumulator')
+          && typeof options['accumulator'] === 'function') {
+        return this.map((user, index, array) => {
+          return options.accumulator(user, index, array);
+        }).calculateAverage();
+      }
+
+      const filteredObjects = this.filter((obj) => {
+        return typeof obj[options['propertyName']] === 'number';
+      });
+
+      if (filteredObjects.length !== 0) {
+        return filteredObjects.map((obj) => {
+          return obj[options['propertyName']];
+        }).calculateAverage();
```

Comment:
filter, map, calcAverage(reduce inside) - has O(3n) complexity, and can be rewritten with 1 reduce - O(n)
Yes, 3 - only constant, it will be 3 times faster